### DISCUSSION TITLE
OpenAI asynchronous support Guidance reborn

### DISCUSSION THREAD

**Discussion by Nick-tenBroeke** [guidance-ai/guidance#478] [Q&A]
Hi Guidance contributors and community,

I'm wanting to make the change to the new Guidance 0.1 but I wanted to verify first whether I can run chat based OpenAI model asynchronously in the new paradigm? The old Guidance version allowed for an async_mode parameter in the program definition but when I try and search in the new source code, there seems to be no implementations of .acreate calls to OpenAI.

It would be of great help to me if someone could either confirm or disconfirm this, and if the former is the case, what would be the minimally viable example implementation of this pattern?

Thanks a lot in advance and thank you for the great work on Guidance!
Nick

**Comment by wiiiktor** [comment]
I use asyncio for a similar task, wouldn't this work with guidance?: 

`
    import asyncio
    templates_1 = ['prompt1 ...', prompt2' ..., 'prompt3 example']
    tab = ['' for i in range(len(memory))]
    tasks_1 = []
    async def generate_concurrently(templates_1, tok):
        for i, t in enumerate(templates_1):
            llm = ChatOpenAI(temperature=0.5, model_name='gpt-3.5-turbo-1106', max_retries=5, request_timeout=44)

            async def async_generate(llm, i, t, tok):
                try:
                    with get_openai_callback() as cb:
                        resp = await llm.ainvoke(t.strip())
                        tab[i] = str(resp)
                        tok['total'] += cb.total_tokens
                except Exception as er:
                    print('error1', er)
            tasks_1.append(async_generate(llm, i, t, tok))
        print('len_tasks_1:', len(tasks_1))
        await asyncio.gather(*tasks_1)

    async def func_1(templates_1, tok):
        try:
            await asyncio.wait_for(generate_concurrently(templates_1, tok), timeout=35)
        except asyncio.TimeoutError as er:
            print('error2timeout', er)
            # print("Not all the tab entries have been generated.\n")

    asyncio.run(func_1(templates_1, tok))
    print('tokens_used:', tok['total'])`

**Reply by Nick-tenBroeke** [reply]
Hi @wiiiktor, thanks for your reply! I've implemented a similar pattern using the old handlebars version of guidance, where the Program object has async_mode set to True, but when I try to to implement this pattern with the new 0.1 paradigm, I'm unable to get it to work. 

This is what I've tried as minimal example, but it throws a TypeError:

`from guidance import user, system, assistant, gen, models
import guidance
from openai import AsyncOpenAI
import asyncio

gpt35 = models.OpenAI("gpt-3.5-turbo", api_key=KEY)
gpt35.client = AsyncOpenAI(api_key=KEY) # overwrite the client with an async one

@guidance
async def experts(lm, query):
    with system():
        lm += "You are a helpful assistant."

    with user():
        lm += f"""\
        I want a response to the following question:
        {query}
        Who are 3 world-class experts (past or present) who would be great at answering this?
        Please don't answer the question or comment on it yet."""

    with assistant():
        lm += gen(name='experts', max_tokens=300)
    
    with user():
        lm += f"""\
        Great, now please answer the question as if these experts had collaborated in writing a joint anonymous answer.
        In other words, their identity is not revealed, nor is the fact that there is a panel of experts answering the question.
        If the experts would disagree, just present their different positions as alternatives in the answer itself (e.g. 'some might argue... others might argue...').
        Please start your answer with ANSWER:"""
    
    with assistant():
        lm += gen(name='answer', max_tokens=500)
   
async def arun_experts():
   return await experts(gpt35, "What is the meaning of life?")

async def run_async_program(timeout):
    task = arun_experts()
    
    try:
        res = await asyncio.wait_for(task, timeout)
        return res
    except asyncio.TimeoutError:
        print("Timeout occurred")
        raise
    
asyncio.run(run_async_program(2))`

**Comment by wiiiktor** [comment]
I will face this problem (asynchronous requests) within 2-3 days, so I will put my results here (if not, you can send me a reminder in this thread)

**Comment by slundberg** [comment]
Hi! We don't ever call the async openai `.acreate` method because we run the openai calls (or any other remote endpoint) in a separate thread, so we don't need async to avoid blocking. I will have to think if there is any reason we want to expose an async api on top of that...if you think of a reason let me know.

What we do really need, but have not finished yet is the streaming api from the 0.0.64 version where you can get a partial state as the program executes. That is on the list of things to add :)

**Reply by Nick-tenBroeke** [reply]
Hi @slundberg, thank you for your reply. I'm using an async wrapper around my guidance programs (.64 version) so that I can decide when I want to abandon them (for instance when their execution is taking too long). I guess I can do this in the new paradigm by joining the separate thread that was created for the openai call. Would that be a correct approach and if so, how can I best join the running thread so that I can terminate it if required? I'm finding it difficult to reason over the new paradigm so any help and code examples would be greatly appreciated!

**Reply by Nick-tenBroeke** [reply]
@slundberg, would be great to still get a reply on this as I would love to move to the new guidance implementation. What I need for my use case is to be able to monitor a running guidance program in a non-blocking way and to be able to forcefully terminate it after a time-out period. In the old version I can do that with this async wrapper that makes the guidance calls but that's no longer possible:

````
async def _run_async_program(self, guidance_func: guidance.Program, func_args, timeout, test_vars: [str]):
        task = guidance_func(**func_args)
        
        try:
            res = await asyncio.wait_for(task, timeout)
            for t in test_vars:
                if t not in res:
                    print(f"{t} not in res")
                    raise TimeoutError(f"Required variable {t} missing in response.")
                elif res._execute_complete._value == False:
                    print("Did not finish executing completely")
                    raise TimeoutError(f"Res object not in _execute_complete state.")
            return res
        except asyncio.TimeoutError:
            print("Timeout occurred")
            await task
            raise
        except Exception as e:
            print(f"An unhandled exception occurred: {e}")
            await task
            raise
````

 Would greatly appreciate a follow up!

**Reply by wille-42** [reply]
> What we do really need, but have not finished yet is the streaming api from the 0.0.64 version where you can get a partial state as the program executes. That is on the list of things to add :)

Hi @slundberg, we're looking forward to the streaming feature being delivered soon. Currently, we run all Guidance v0.0.64 programs/functions with `async_mode=True` and `stream=True` using `asyncio`. We hope the new API will support us in obtaining partial program outputs not only in a streaming manner but also asynchronously.

**Comment by wiiiktor** [comment]
So, if I want to run 100 requests at the same time, how to I do it? Do you please have a code sample for this / a small manual?
